cores/esp8266/core_esp8266_i2s.c

```diff
@@ -30,147 +30,278 @@
+    ETS_SLC_INTR_DISABLE();
+    rx_irqs++;
+    slc_queue_item_t *finished_item = (slc_queue_item_t *)SLCTXEDA;
+    finished_item->owner = 1; // Or else RX just stops
+    finished_item->myid++;
+    i2s_slc_queue_append_item(rx, finished_item->buf_ptr);
+    if (rx->callback) rx->callback();
     ETS_SLC_INTR_ENABLE();
   }
 }
 
 void i2s_set_callback(void (*callback) (void)){
-    i2s_callback = callback;
+  tx->callback = callback;
+}
+
+void i2s_rx_set_callback(void (*callback) (void)){
+  rx->callback = callback;
 }
 
-void ICACHE_FLASH_ATTR i2s_slc_begin(){
-  i2s_slc_queue_len = 0;
+static void ICACHE_FLASH_ATTR _alloc_channel(i2s_state_t *ch) {
   int x, y;
-  
+
+  ch->slc_queue_len = 0;
   for (x=0; x<SLC_BUF_CNT; x++) {
-    i2s_slc_buf_pntr[x] = malloc(SLC_BUF_LEN*4);
-    for (y=0; y<SLC_BUF_LEN; y++) i2s_slc_buf_pntr[x][y] = 0;
-
-    i2s_slc_items[x].unused = 0;
-    i2s_slc_items[x].owner = 1;
-    i2s_slc_items[x].eof = 1;
-    i2s_slc_items[x].sub_sof = 0;
-    i2s_slc_items[x].datalen = SLC_BUF_LEN*4;
-    i2s_slc_items[x].blocksize = SLC_BUF_LEN*4;
-    i2s_slc_items[x].buf_ptr = (uint32_t)&i2s_slc_buf_pntr[x][0];
-    i2s_slc_items[x].next_link_ptr = (int)((x<(SLC_BUF_CNT-1))?(&i2s_slc_items[x+1]):(&i2s_slc_items[0]));
+    ch->slc_buf_pntr[x] = malloc(SLC_BUF_LEN*4);
+    for (y=0; y<SLC_BUF_LEN; y++) ch->slc_buf_pntr[x][y] = 0;
+
+    ch->slc_items[x].unused = 0;
+    ch->slc_items[x].owner = 1;
+    ch->slc_items[x].eof = 1;
+    ch->slc_items[x].sub_sof = 0;
+    ch->slc_items[x].datalen = SLC_BUF_LEN * 4;
+    ch->slc_items[x].blocksize = SLC_BUF_LEN * 4;
+    ch->slc_items[x].buf_ptr = (uint32_t*)&ch->slc_buf_pntr[x][0];
+    ch->slc_items[x].next_link_ptr = (x<(SLC_BUF_CNT-1))?(&ch->slc_items[x+1]):(&ch->slc_items[0]);
+    ch->slc_items[x].myid = 0;
+  }
+}
+#if 0
+void dumprx()
+{
+for (int i=0; i<SLC_BUF_CNT; i++) {
+printf("%d: %d %d %d %d %d %d %p %p\n", i, rx->slc_items[i].myid, rx->slc_items[i].owner, rx->slc_items[i].eof, rx->slc_items[i].sub_sof, rx->slc_items[i].datalen, rx->slc_items[i].blocksize,
+rx->slc_items[i].buf_ptr, rx->slc_items[i].next_link_ptr);
+}
+}
+#endif
+
+
+static void ICACHE_FLASH_ATTR i2s_slc_begin() {
+  if (tx) {
+    _alloc_channel(tx);
+  }
+  if (rx) {
+    _alloc_channel(rx);
   }
 
   ETS_SLC_INTR_DISABLE();
   SLCC0 |= SLCRXLR | SLCTXLR;
   SLCC0 &= ~(SLCRXLR | SLCTXLR);
   SLCIC = 0xFFFFFFFF;
 
-  //Configure DMA
-  SLCC0 &= ~(SLCMM << SLCM); //clear DMA MODE
-  SLCC0 |= (1 << SLCM); //set DMA MODE to 1
-  SLCRXDC |= SLCBINR | SLCBTNR; //enable INFOR_NO_REPLACE and TOKEN_NO_REPLACE
-  SLCRXDC &= ~(SLCBRXFE | SLCBRXEM | SLCBRXFM); //disable RX_FILL, RX_EOF_MODE and RX_FILL_MODE
+  // Configure DMA
+  SLCC0 &= ~(SLCMM << SLCM); // Clear DMA MODE
+  SLCC0 |= (1 << SLCM); // Set DMA MODE to 1
+  SLCRXDC |= SLCBINR | SLCBTNR; // Enable INFOR_NO_REPLACE and TOKEN_NO_REPLACE
+  SLCRXDC &= ~(/*SLCBRXFE |*/ SLCBRXEM | SLCBRXFM); //disable RX_FILL, RX_EOF_MODE and RX_FILL_MODE
```

Comment:
if RX_FILL is no longer disabled, then the comment should be updated

Reply:
Still trying to figure out *why* it's working the way it is.  No complete register map AFAICT for this monster.  Actually two monsters.  The SLC is partially documented, the I2S is almost a black box...